libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,447 @@
+    std::vector<polynomial<libff::Fr<ppT>>> Q_polys,
+    std::vector<libff::G1<ppT>> &Q_polys_at_secret_g1)
+{
+    assert(secret_powers_g1.size());
+    assert(Q_polys.size());
+    assert(Q_polys.size() <= secret_powers_g1.size());
+    const size_t npolys = Q_polys.size();
+    for (size_t i = 0; i < npolys; ++i) {
+        Q_polys_at_secret_g1[i] =
+            plonk_evaluate_poly_at_secret_G1<ppT>(secret_powers_g1, Q_polys[i]);
+    }
+}
+
+// Compute the factors in the product of the permutation polynomial
+// z(X) in Prover Round 2. Note that accumulator A[0]=1 and A[i],
+// i>0 is computed from values at i-1 for witness[i-1], H_gen[i-1],
+// H_gen_permute[i-1]m etc.
+template<typename FieldT>
+FieldT plonk_compute_accumulator_factor(
+    size_t i,
+    size_t n, // nconstraimts
+    FieldT beta,
+    FieldT gamma,
+    std::vector<FieldT> witness,
+    std::vector<FieldT> H_gen, // H, Hk1, Hk2
+    std::vector<FieldT> H_gen_permute,
+    std::vector<FieldT> &A // accumulatro vector
+)
+{
+    assert(n);
+    assert((i >= 0) && (i < n));
+    assert(witness.size() == (3 * n));
+    assert(H_gen.size() == (3 * n));
+    assert(H_gen_permute.size() == (3 * n));
+    assert(A.size() == n);
+    FieldT res = FieldT(1);
+    if (i > 0) {
+        FieldT nom_1 = witness[i - 1] + (beta * H_gen[i - 1]) + gamma;
+        FieldT den_1 = witness[i - 1] + (beta * H_gen_permute[i - 1]) + gamma;
+
+        FieldT nom_2 = witness[n + i - 1] + (beta * H_gen[n + i - 1]) + gamma;
+        FieldT den_2 =
+            witness[n + i - 1] + (beta * H_gen_permute[n + i - 1]) + gamma;
+
+        FieldT nom_3 =
+            witness[2 * n + i - 1] + (beta * H_gen[2 * n + i - 1]) + gamma;
+        FieldT den_3 = witness[2 * n + i - 1] +
+                       (beta * H_gen_permute[2 * n + i - 1]) + gamma;
+
+        FieldT nom = nom_1 * nom_2 * nom_3;
+        FieldT den = den_1 * den_2 * den_3;
+
+        res = nom * den.inverse() * A[i - 1];
+    }
+    return res;
+}
+
+template<typename FieldT>
+void plonk_compute_accumulator(
+    size_t n, // num_gates
+    FieldT beta,
+    FieldT gamma,
+    std::vector<FieldT> witness,
+    std::vector<FieldT> H_gen, // H, Hk1, Hk2
+    std::vector<FieldT> H_gen_permute,
+    std::vector<FieldT> &A // accumulatro vector
+)
+{
+    assert(n);
+    assert(witness.size() == (3 * n));
+    assert(H_gen.size() == (3 * n));
+    assert(H_gen_permute.size() == (3 * n));
+    assert(A.size() == n);
+    for (size_t i = 0; i < n; ++i) {
+        A[i] = plonk_compute_accumulator_factor(
+            i, n, beta, gamma, witness, H_gen, H_gen_permute, A);
+    }
+}
+
+//
+// Note: the following function was copied from
+// libff/algebra/curves/tests/test_groups.cpp TODO: maybe add the
+// declaration in a header to use it directly from here. Uded in the
+// Verifier code.
+//
+// Check that a group element satisfies the curve equation Y^2 = X^3
+// + a X + b
+//
+template<typename GroupT> bool check_curve_equation(GroupT P)
```

Comment:
As discussed, there should be a libff function to do this.

Reply:
Function not used. Removed.